### ISSUE TITLE
Error handler callback has return type bool but allows others

### ISSUE THREAD

**Issue by othercorey** [php/doc-en#1206]
https://www.php.net/manual/en/function.set-error-handler.php

The `handler` callback passed to `set_error_handler` is documented with return type `bool`. The description for the callback includes:

> If the function returns false then the normal error handler continues.

However, this function can return more than just boolean values. It seems to ignore anything returned except `false`.

I'm not sure this is an actual bug, but I wanted to clarify that it can be confusing when discussing complicated error handler flows where some developers returned boolean values, some returned nothing and so on.

Will the callback ever error on non-boolean values? Should we explain that any other value has no effect?

**Comment by damianwadley** [comment]
The behavior is literally what it says there: if it returns `false`, not 0 or an empty string or anything _loosely_ false, then the normal error handler is invoked. PHP doesn't care whether it returns any other value or even no value at all.

```php
if ($callback($errno, $errstr, $errfile, $errline) === false) {
    normal_error_handler($errno, $errstr, $errfile, $errline);
}
```

**Comment by othercorey** [comment]
You've simply restated my point.

**Comment by damianwadley** [comment]
I'll try again with some quotes:

> I'm not sure this is an actual bug,

It's not.

> it can be confusing when discussing complicated error handler flows where some developers returned boolean values, some returned nothing and so on

What's confusing about "if it returns false then the normal error handler continues"? If you return false then it does that, and if you don't return false then it doesn't do that.

> Will the callback ever error on non-boolean values?

"Ever"? Maybe something will happen in PHP 9 or 10 or 15. But for now, no: all it cares about is false.

> Should we explain that any other value has no effect?

I thought it was implied, but I also haven't closed this issue.

**Comment by othercorey** [comment]
Why are you being so aggressive?

I'm asking about a simple documentation improvement for something that is a little contradictory.

**Comment by ADmad** [comment]
Given that PHP now has stricter typing, one wouldn't be at fault thinking a `bool` means the function can/should only return boolean values. While the description provides future details the `: bool` for the callback is pretty misleading.

**Comment by ADmad** [comment]
> PHP doesn't care whether it returns any other value or even no value at all.

Aren't return types meant to indicate what a function can return? What PHP does with the return value isn't the issue here. Having `bool` return type is misleading since the callback can return a non-boolean value and PHP will act accordingly.

**Comment by ADmad** [comment]
Perhaps in a future minor returning any other value besides `false` from the callback should be deprecated.

**Comment by cmb69** [comment]
Well, this is a bit tricky, since a user defined error handler is called from the Zend engine, and as such behaves like it was called from a file without declare(strict_types=1). So if you set an error handler without return type declaration, you can return any value from it, but only `false` values have special meaning. If you declare the return as `bool`, however, the actual return value is coerced to bool, so returning any falsy value will actually return `false`.

This is standard behavior, but of course depends on the actual declaration of the handler function. Maybe we should just state that the error handler *should* have that signature, but users are free to omit the type declarations?

**Comment by ADmad** [comment]
> the actual return value is coerced to bool, so returning any falsy value will actually return false.

This contradicts what @damianwadley said in his 1st comment: *"The behavior is literally what it says there: if it returns false, not 0 or an empty string or anything loosely false, then the normal error handler is invoked."*

Not trying to point any fingers here, just highlighting the fact how the existing docs are inadequate.

If the return value is indeed coerced to boolean than the `bool` return type seems reasonable. This fact should be added to the description.

**Comment by ADmad** [comment]
Sorry, I misunderstood, the coercion is done *only* if the callback has return type `bool`, not if type is omitted.

**Comment by othercorey** [comment]
Ok, so it's good to know there is standard behavior here. When the docs state there is action on `false`, that means an `=== false` check, but the return value could be coerced to true or false depending on the type declaration.

This means the existing code that does just `return;` would be coerced to `false` if the correct return type `bool` is added. 

So, there is a change of behavior when adding type declarations? That is the confusion I was hoping to clarify.

**Comment by cmb69** [comment]
Yes, this is just like with userland code in coercive mode (i.e. without `declare(strict_types=1)`), e.g. https://3v4l.org/v2FW1.

**Comment by othercorey** [comment]
In that case, I don't know that an explanation in this doc is the right place unless it doesn't happen with other callbacks.

If there is an explanation of strict types and coercion for callbacks executed by the zend engine, maybe it could just link to it.

**Comment by cmb69** [comment]
Well, https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.strict has some info about coercive mode, but that still is not explicit about that you can just omit the return type declaration of the error handler. Also note that other callback functions have similar behavior, e.g. an [Sqlite3::createCollation()](https://www.php.net/manual/en/sqlite3.createcollation.php) callback [is supposed to return an int, but if it doesn't, there's a warning](https://github.com/php/php-src/blob/192ea91debe10ba23f1576f43a666f5947ac3943/ext/sqlite3/sqlite3.c#L933-L940); that won't happen, if the handler has a respective return type declaration. It seems to be a good idea to enforce suitable return type declarations, but that would be a BC break.

**Comment by ADmad** [comment]
Thank you for the explanation @cmb69. 

A warning when an invalid type is returned, like it's done for `Sqlite3::createCollation()`'s callback, would be quite nice. Adding similar warning for `set_error_handler()`'s callback in the next minor release (8.2) would be within the BC guidelines right?

**Comment by othercorey** [comment]
@cmb69 I would help write a section on this, but I'm afraid it's going to be from a naïve understanding of the internals.
